GitOut/Features/Git/Log/GitTreeEvent.cs

```diff
@@ -0,0 +1,154 @@
+            new AvailableColor(Color.FromRgb(200, 200, 200)),
+            new AvailableColor(Color.FromRgb(100, 100, 100)),
+            new AvailableColor(Color.FromRgb(50, 50, 50))
+        };
+
+        private readonly List<GitTreeNode> nodes = new List<GitTreeNode>();
+
+        public GitTreeEvent(GitHistoryEvent historyEvent) => Event = historyEvent;
+
+        public int CommitIndex { get; set; } = -1;
+        public GitHistoryEvent Event { get; }
+        public IReadOnlyCollection<GitTreeNode> Nodes => nodes;
+
+        public IEnumerable<TreeBuildingLeaf> Process(IEnumerable<TreeBuildingLeaf> leafs) => ProcessBottom(ProcessTop(leafs));
+
+        public static void BeginProcessing() => colors.ForEach(c => c.Available = true);
+
+        private IEnumerable<TreeBuildingLeaf> ProcessBottom(IEnumerable<TreeBuildingLeaf> leafs)
+        {
+            int from = 0;
+            int to = 0;
+            var bottomLeafs = new List<TreeBuildingLeaf>();
+            bool processedCommit = false;
+            foreach (TreeBuildingLeaf leaf in leafs)
+            {
+                if (from == CommitIndex)
+                {
+                    processedCommit = true;
+
+                    if (Event.Parent != null)
+                    {
+                        leaf.Current.Bottom = new Line(from, to++);
+                        bottomLeafs.Add(TreeBuildingLeaf.WithParent(Event.Parent, leaf.Current));
+                        if (Event.MergedParent != null)
+                        {
+                            var node = GitTreeNode.WithBottomLine(new Line(from, to++), GetNextAvailableColor(), true);
+                            nodes.Add(node);
+                            bottomLeafs.Add(TreeBuildingLeaf.WithParent(Event.MergedParent, node));
+                        }
+                    }
+                    ++from;
+                }
+                else
+                {
+                    leaf.Current.Bottom = new Line(from++, to++);
+                    bottomLeafs.Add(TreeBuildingLeaf.WithParent(leaf.LookingFor!, leaf.Current));
+                }
+            }
+            if (!processedCommit)
+            {
+                var node = GitTreeNode.WithBottomLine(new Line(from, to), GetNextAvailableColor(), true);
+                nodes.Add(node);
+                if (Event.Parent != null)
+                {
+                    bottomLeafs.Add(TreeBuildingLeaf.WithParent(Event.Parent, node));
+                    if (Event.MergedParent != null)
+                    {
+                        var mergedNode = GitTreeNode.WithBottomLine(new Line(from, to++), GetNextAvailableColor(), true);
+                        nodes.Add(mergedNode);
+                        bottomLeafs.Add(TreeBuildingLeaf.WithParent(Event.MergedParent, mergedNode));
+                    }
+                }
+            }
+            return bottomLeafs;
+        }
+
+        private Color GetNextAvailableColor()
```

Comment:
Maybe move this and SetColorAvailable to bottom so that ProcessBottom and ProcessTop are together